From 2eb1380255e0ba5148b35551421077da9d7c2306 Mon Sep 17 00:00:00 2001
From: nomoto <nomoto@aptpod.co.jp>
Date: Tue, 7 Jul 2026 17:38:38 +0900
Subject: [PATCH] fix(netdev): assign netdev before register_candev to avoid
 probe race

Backport of cdf578c to 1.7.0 (AP-CT2A / EP1-CH02A).
---
 module/ap_ct2a/ap_ct2a_fops.c     | 11 ++++++-----
 module/ap_ct2a/ap_ct2a_main.c     |  5 +++--
 module/ep1_ch02a/ep1_ch02a_main.c |  5 +++--
 3 files changed, 12 insertions(+), 9 deletions(-)

diff --git a/module/ap_ct2a/ap_ct2a_fops.c b/module/ap_ct2a/ap_ct2a_fops.c
index 08df00f..0c0cff1 100644
--- a/module/ap_ct2a/ap_ct2a_fops.c
+++ b/module/ap_ct2a/ap_ct2a_fops.c
@@ -378,13 +378,14 @@ int apt_usbtrx_unique_can_netdev_set_bittiming(struct net_device *netdev)
 /*!
  * @brief start interface
  */
-static int apt_usbtrx_unique_can_netdev_start(apt_usbtrx_dev_t *dev)
+static int apt_usbtrx_unique_can_netdev_start(struct net_device *netdev)
 {
+	apt_usbtrx_candev_t *candev = netdev_priv(netdev);
+	apt_usbtrx_dev_t *dev = candev->dev;
 	apt_usbtrx_unique_data_can_t *unique_data = get_unique_data(dev);
-	apt_usbtrx_candev_t *candev = netdev_priv(unique_data->netdev);
 	int result;
 
-	result = candev->can.do_set_bittiming(unique_data->netdev);
+	result = candev->can.do_set_bittiming(netdev);
 	if (result != 0) {
 		EMSG("can.do_set_bittiming().. Error");
 		return -EIO;
@@ -426,7 +427,7 @@ int apt_usbtrx_unique_can_netdev_open(struct net_device *netdev)
 	}
 
 	/* finally start device */
-	err = apt_usbtrx_unique_can_netdev_start(dev);
+	err = apt_usbtrx_unique_can_netdev_start(netdev);
 	if (err) {
 		netdev_err(netdev, "couldn't start device: %d\n", err);
 		close_candev(netdev);
@@ -534,7 +535,7 @@ int apt_usbtrx_unique_can_netdev_set_mode(struct net_device *netdev, enum can_mo
 			return -EIO;
 		}
 
-		result = apt_usbtrx_unique_can_netdev_start(dev);
+		result = apt_usbtrx_unique_can_netdev_start(netdev);
 		if (result) {
 			netdev_err(netdev, "couldn't start device: %d\n", result);
 			return result;
diff --git a/module/ap_ct2a/ap_ct2a_main.c b/module/ap_ct2a/ap_ct2a_main.c
index 8456ef4..957891e 100644
--- a/module/ap_ct2a/ap_ct2a_main.c
+++ b/module/ap_ct2a/ap_ct2a_main.c
@@ -76,15 +76,16 @@ static int apt_usbtrx_unique_can_create_candev(struct usb_interface *intf, const
 	SET_NETDEV_DEV(netdev, &intf->dev);
 	netdev->dev_id = dev->ch;
 
+	unique_data->netdev = netdev;
+
 	err = register_candev(netdev);
 	if (err) {
+		unique_data->netdev = NULL;
 		free_candev(netdev);
 		EMSG("register_candev().. Error");
 		return err;
 	}
 
-	unique_data->netdev = netdev;
-
 	netdev_info(netdev, "device %s registered\n", netdev->name);
 
 	return 0;
diff --git a/module/ep1_ch02a/ep1_ch02a_main.c b/module/ep1_ch02a/ep1_ch02a_main.c
index 8bc2745..3af5e24 100644
--- a/module/ep1_ch02a/ep1_ch02a_main.c
+++ b/module/ep1_ch02a/ep1_ch02a_main.c
@@ -118,15 +118,16 @@ static int ep1_ch02a_create_candev(struct usb_interface *intf, const struct usb_
 	SET_NETDEV_DEV(netdev, &intf->dev);
 	netdev->dev_id = dev->ch;
 
+	unique_data->netdev = netdev;
+
 	err = register_candev(netdev);
 	if (err) {
+		unique_data->netdev = NULL;
 		free_candev(netdev);
 		EMSG("register_candev().. Error");
 		return err;
 	}
 
-	unique_data->netdev = netdev;
-
 	netdev_info(netdev, "device %s registered\n", netdev->name);
 
 	return 0;
-- 
2.43.0

